File: day9b.py
#!/usr/bin/env python3

plen = 25
with open("day9.txt") as f:
    data = f.read()
data = list(map(int, data.splitlines()))

def check(val, preamble):
    return True in [(x != y) and (x + y) == val for x in preamble for y in preamble]

for i in range(plen, len(data)):
    if not check(data[i], data[i - plen:i]):
        found_i = i
        found_val = data[i]
        #print(found_i, found_val)
        break

candidates = data[:found_i]
for c_s in range(len(candidates)):
    for c_e in range(c_s, len(candidates)):
        l = candidates[c_s:c_e]
        if sum(l) == found_val:
            print(min(l) + max(l))
            break

File: day9b.hs
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

plen :: Int
plen = 25

parse :: T.Text -> [Int]
parse content = map (read . T.unpack) $ T.lines $ content

isValid :: [Int] -> Int -> Bool
isValid prefix value =
    or [True | x <- prefix, y <- prefix, x /= y, x + y == value]

findInvalid :: [Int] -> [Int] -> Int -> (Int, Int)
findInvalid prefix (value:values) i =
    if not $ isValid prefix value then
        (i, value)
    else
        findInvalid ((tail prefix) ++ [value]) values (i + 1)

findSumTo :: [Int] -> Int -> [Int]
findSumTo vals val = head validLists
    where
        nvals = length vals
        lists = [take n $ drop i vals | n <- [2..nvals], i <- [2..(nvals - n)]]
        validLists = filter (\l -> sum l == val) lists

answer :: [Int] -> Int
answer content = sumToMin + sumToMax
    where
        prefix = take plen content
        values = drop plen content
        (invalidPos, invalidVal) = findInvalid prefix values 0
        sumTo = findSumTo (take (plen + invalidPos) content) invalidVal
        sumToMin = minimum sumTo
        sumToMax = maximum sumTo

main :: IO ()
main = do
    content <- TIO.readFile "day9.txt"
    let input = parse content
    putStrLn $ show $ answer input
